Bugfix/Fix LPADC MCXN947 test case failure

ZhaoxiangJin opened the pull request:
Change method to get adc ref voltage. When an application enables the USERSPACE, the driver cannot obtain the correct vref_mv by using 'CONTAINER_OF', because the z_vrf_adc_channel_setup function will not copy vref_mv from user space to kernel space.

Fixes #74842

butok commented:
What does the number "74842 bug" mean?

ZhaoxiangJin commented:
> What does the number "74842 bug" mean?

https://github.com/zephyrproject-rtos/zephyr/issues/74842

dleach02 commented:
@ZhaoxiangJin, it is enough to just use a title like:

"Fix LPADC MCXN947 test case failure"

And in the body of the commit message use:

"Fixes #74842"

This will allow the Github automatic linking to occur so that when this PR merges, the 74842 issue will automatically close.

dleach02 commented on samples/drivers/adc/adc_dt/sample.yaml:
```diff
@@ -32,6 +32,7 @@ tests:
       - longan_nano
```
you can combine your testscope changes and the overlay into a single commit

ZhaoxiangJin replied:
ok, I will combine them, thanks for your suggestions.

decsny replied:
yes, please combine all the test changes to one commit since they are so simple

dleach02 commented on dts/bindings/adc/nxp,lpc-lpadc.yaml:
```diff
@@ -75,6 +75,12 @@ properties:
     type: phandles
```
look at the history on the file for some hints on the syntax to use. 

dts: bindings: lpadc: Add new ref-supply-val property

ZhaoxiangJin replied:
ok, I will update the commit message, make it more syntax-compliant.

decsny commented on dts/bindings/adc/nxp,lpc-lpadc.yaml:
```diff
@@ -75,6 +75,12 @@ properties:
     type: phandles
     description: References to required regulators which must be enabled for LPADC to function
 
+  ref-supply-val:
```
instead of adding a new property, maybe we can make the nxp,reference-supply a phandle array?

ZhaoxiangJin committed:
dts: bindings: vref: Add specifier cells for nxp vref

Add specifier cells for nxp vref, when some other
peripherals use the vref to provide reference voltage,
this cells can be used to pass-in vref target voltage.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: bindings: lpadc: Change 'nxp,reference-supply' property

Remove 'nxp,reference-supply' property.
Add new phandle-array type 'nxp,references' property.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: arm: nxp_mcxn947: Add 'nxp,reference-cells' for vref node

Add new property 'nxp,reference-cells' for vref node.
Remove lpadc nodes 'nxp,reference-supply' property.
Add new property 'nxp,references' for lpadc nodes.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: lpadc: Update lpadc driver to use the latest feature

Updated nxp lpadc driver, because the phandle type property
'nxp,reference-supply' was removed, and a new phandle-array
type property 'nxp,references' added.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: Add frdm_mcxn947 into regulator and adc test scope

Add frdm_mcxn947 into regulator and adc test scope

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: adc: Add 'zephyr,vref-mv' property.

Add 'zephyr,vref-mv' property to frdm_mcxn947 overlay file.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: frdm_mcxn947: Update index.rst

Update index.rst, lowercase regulator.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on dts/bindings/adc/nxp,lpc-lpadc.yaml:
```diff
@@ -71,8 +71,8 @@ properties:
     required: true
     description: Offset value B to use if CONFIG_LPADC_DO_OFFSET_CALIBRATION is false
 
-  nxp,reference-supply:
```
I renamed this property and its type (may also can be considered as deprecated and add a new property), should I add this information to the doc/releases/release-notes-3.7.rst? Seems 3.7 has already been tagged.

ZhaoxiangJin committed:
docs: migration-guide: Add NXP LPADC driver update info

Add NXP LPADC (nxp,lpc-lpadc) driver update info.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

decsny commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -241,6 +241,12 @@ Device Drivers and Devicetree
 Analog-to-Digital Converter (ADC)
 =================================
 
+* The DT property ``nxp,reference-supply`` of :dtcompatible:`nxp,lpc-lpadc` driver has
```
this would be better to go in the DT binding section

decsny commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -269,6 +269,10 @@ Drivers and Sensors
 
+  * The NXP LPADC driver now supports passing the reference voltage value by using new
+    phandle-array type DT property ``nxp,references``. The DT property ``nxp,reference-supply`` 
+    is now deprecated.
```
don't say "deprecated" because that means the old thing is still around, we just changed this. also, keep this in this ADC section unlike the migration guide change

dleach02 commented:
@ZhaoxiangJin I've pulled this out of draft mode. Please review and address the compliance check errors.

ZhaoxiangJin commented:
> @ZhaoxiangJin I've pulled this out of draft mode. Please review and address the compliance check errors.

I am currently out of office, I will fix compliance check errors and review comments immediately when I come back.

hakehuang commented:
test pass
```
2024-07-04 13:32:08,937 - twister - DEBUG - DEVICE: TESTSUITE adc_basic succeeded
2024-07-04 13:32:08,938 - twister - DEBUG - DEVICE: 
2024-07-04 13:32:08,940 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY START ------
2024-07-04 13:32:08,941 - twister - DEBUG - DEVICE: 
2024-07-04 13:32:08,950 - twister - DEBUG - DEVICE: SUITE PASS - 100.00% [adc_basic]: pass = 6, fail = 0, skip = 0, total = 6 duration = 0.533 seconds
2024-07-04 13:32:08,955 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_asynchronous_call] duration = 0.006 seconds
2024-07-04 13:32:08,962 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_invalid_request] duration = 0.012 seconds
2024-07-04 13:32:08,969 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_repeated_samplings] duration = 0.091 seconds
2024-07-04 13:32:08,975 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_one_channel] duration = 0.007 seconds
2024-07-04 13:32:08,983 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_two_channels] duration = 0.007 seconds
2024-07-04 13:32:08,988 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_with_interval] duration = 0.410 seconds
2024-07-04 13:32:08,989 - twister - DEBUG - DEVICE: 
2024-07-04 13:32:08,993 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY END ------
2024-07-04 13:32:08,993 - twister - DEBUG - DEVICE: 
2024-07-04 13:32:08,998 - twister - DEBUG - DEVICE: ===================================================================
2024-07-04 13:32:09,002 - twister - DEBUG - DEVICE: RunID: deaa3cd0d691bd6a721b4be926b46bdb
2024-07-04 13:32:09,005 - twister - DEBUG - DEVICE: PROJECT EXECUTION SUCCESSFUL
```

ZhaoxiangJin commented:
> test pass
> 
> ```
> 2024-07-04 13:32:08,937 - twister - DEBUG - DEVICE: TESTSUITE adc_basic succeeded
> 2024-07-04 13:32:08,938 - twister - DEBUG - DEVICE: 
> 2024-07-04 13:32:08,940 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY START ------
> 2024-07-04 13:32:08,941 - twister - DEBUG - DEVICE: 
> 2024-07-04 13:32:08,950 - twister - DEBUG - DEVICE: SUITE PASS - 100.00% [adc_basic]: pass = 6, fail = 0, skip = 0, total = 6 duration = 0.533 seconds
> 2024-07-04 13:32:08,955 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_asynchronous_call] duration = 0.006 seconds
> 2024-07-04 13:32:08,962 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_invalid_request] duration = 0.012 seconds
> 2024-07-04 13:32:08,969 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_repeated_samplings] duration = 0.091 seconds
> 2024-07-04 13:32:08,975 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_one_channel] duration = 0.007 seconds
> 2024-07-04 13:32:08,983 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_two_channels] duration = 0.007 seconds
> 2024-07-04 13:32:08,988 - twister - DEBUG - DEVICE: - PASS - [adc_basic.test_adc_sample_with_interval] duration = 0.410 seconds
> 2024-07-04 13:32:08,989 - twister - DEBUG - DEVICE: 
> 2024-07-04 13:32:08,993 - twister - DEBUG - DEVICE: ------ TESTSUITE SUMMARY END ------
> 2024-07-04 13:32:08,993 - twister - DEBUG - DEVICE: 
> 2024-07-04 13:32:08,998 - twister - DEBUG - DEVICE: ===================================================================
> 2024-07-04 13:32:09,002 - twister - DEBUG - DEVICE: RunID: deaa3cd0d691bd6a721b4be926b46bdb
> 2024-07-04 13:32:09,005 - twister - DEBUG - DEVICE: PROJECT EXECUTION SUCCESSFUL
> ```

Thanks for your work.

dleach02 commented:
@ZhaoxiangJin please address the merge conflict

ZhaoxiangJin commented:
> @ZhaoxiangJin please address the merge conflict

Rebased.

ZhaoxiangJin commented:
Hi @anangl, could you please help to review this PR? Thanks.

anangl reviewed (approved):
nit: commit "drivers: lpadc: Update lpadc driver to use the latest feature" says that the "nxp,reference-supply" property has been deprecated, not removed.

ZhaoxiangJin commented:
> nit: commit "drivers: lpadc: Update lpadc driver to use the latest feature" says that the "nxp,reference-supply" property has been deprecated, not removed.

Yes, this commit message might confuse users, I updated the message, thanks for your review.
